### PR TITLE
Add libsndfile, libflac, libvorbis, libopus

### PR DESCRIPTION
**Pull request by SirLynix**

### CODE REVIEW

**Comment by waruqi** [comment]
Please remove  .vscode/compile_commands.json

File: packages/l/libsndfile/xmake.lua [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+            local vs_runtime = package:config("vs_runtime")
+            if vs_runtime == "MT" then
+                table.insert(configs, "-DCMAKE_MSVC_RUNTIME_LIBRARY=MultiThreaded")
+            elseif vs_runtime == "MTd" then
+                table.insert(configs, "-DCMAKE_MSVC_RUNTIME_LIBRARY=MultiThreadedDebug")
+            elseif vs_runtime == "MD" then
+                table.insert(configs, "-DCMAKE_MSVC_RUNTIME_LIBRARY=MultiThreadedDLL")
+            elseif vs_runtime == "MDd" then
+                table.insert(configs, "-DCMAKE_MSVC_RUNTIME_LIBRARY=MultiThreadedDebugDLL")
+            end
```
Comment (waruqi) [inline packages/l/libsndfile/xmake.lua:28]: import("package.tools.cmake").install will pass vs_runtime flags, it does not work?

File: packages/l/libsndfile/xmake.lua [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+       end
+
+        import("package.tools.cmake").install(package, configs)
+   end)
+
+    on_test(function (package)
+        assert(package:check_cxxsnippets({test = [[
+            void test() {
+                const char* version = sf_version_string();
```
Comment (waruqi) [inline packages/l/libsndfile/xmake.lua:37]: you need only use `package:check_cfuncs`

File: packages/l/libsndfile/xmake.lua (hunk 1 above)
Comment (SirLynix) [reply packages/l/libsndfile/xmake.lua:28]: Oh I didn't notice, I saw that CMake 3.15 added [CMAKE_MSVC_RUNTIME_LIBRARY](https://cmake.org/cmake/help/git-stage/variable/CMAKE_MSVC_RUNTIME_LIBRARY.html) and found no reference to it in xmake source. Didn't know it was using another way.

File: packages/l/libogg/xmake.lua [hunk 3]
```diff
@@ -7,14 +7,12 @@ package("libogg")
              "https://gitlab.xiph.org/xiph/ogg.git")
     add_versions("1.3.4", "fe5670640bd49e828d64d2879c31cb4dde9758681bb664f9bdbf159a01b0c76e")
```
Comment (waruqi) [inline packages/l/libogg/xmake.lua:9]: we need add_deps("cmake")

File: packages/l/libopus/xmake.lua [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+
+    add_versions("1.3.1", "65b58e1e25b2a114157014736a3d9dfeaad8d41be1c8179866f144a2fb44ff9d")
+
+    add_deps("cmake")
+
+    on_install("windows", "linux", "macosx", "iphoneos", "mingw", "android", function (package)
+        local configs = {}
+        table.insert(configs, "-DCMAKE_BUILD_TYPE=" .. (package:debug() and "Debug" or "Release"))
+        table.insert(configs, "-DBUILD_SHARED_LIBS=" .. (package:config("shared") and "ON" or "OFF"))
+
```
Comment (waruqi) [inline packages/l/libopus/xmake.lua:17]: Please remove empty line

File: packages/l/libsndfile/xmake.lua [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+    add_deps("cmake", "libflac", "libogg", "libopus", "libvorbis")
+
+    on_install("windows", "linux", "macosx", "iphoneos", "mingw", "android", function (package)
+        local configs = {}
+        table.insert(configs, "-DBUILD_PROGRAMS=OFF")
+        table.insert(configs, "-DBUILD_EXAMPLES=OFF")
+        table.insert(configs, "-DBUILD_TESTING=OFF")
+        table.insert(configs, "-DCMAKE_BUILD_TYPE=" .. (package:debug() and "Debug" or "Release"))
+        table.insert(configs, "-DBUILD_SHARED_LIBS=" .. (package:config("shared") and "ON" or "OFF"))
+
```
Comment (waruqi) [inline packages/l/libsndfile/xmake.lua:21]: and here

**Comment by SirLynix** [comment]
The cmake file missing seems to be an old bug still not fully resolved: https://github.com/xiph/opus/issues/129

**Comment by waruqi** [comment]
> The cmake file missing seems to be an old bug still not fully resolved: [xiph/opus#129](https://github.com/xiph/opus/issues/129)

maybe we need use `add_patches()` to apply a patch

**Comment by SirLynix** [comment]
```
  Could NOT find Ogg (missing: OGG_LIBRARY OGG_INCLUDE_DIR)
```

Although Ogg was installed before, maybe I have to tell it where to find it?

**Comment by SirLynix** [comment]
I think there's a link order problem on Linux:
```
> /usr/bin/g++ -o /dev/shm/.xmake1001/201212/_070815D3FE9144308463A4D2F4A6F940.b /dev/shm/.xmake1001/201212/_AF56D453AFBF4800833DFEDB0B54E920.o -m64 -L/home/runner/.xmake/packages/l/libsndfile/1.0.30/531f88cff3a64da0aa3fb05ee281ab7d/lib -L/home/runner/.xmake/packages/l/libflac/1.3.3/531f88cff3a64da0aa3fb05ee281ab7d/lib -L/home/runner/.xmake/packages/l/libogg/1.3.4/531f88cff3a64da0aa3fb05ee281ab7d/lib -L/home/runner/.xmake/packages/l/libvorbis/1.3.7/531f88cff3a64da0aa3fb05ee281ab7d/lib -L/home/runner/.xmake/packages/l/libopus/1.3.1/531f88cff3a64da0aa3fb05ee281ab7d/lib -lsndfile -lFLAC -logg -lvorbisenc -lvorbis -lvorbisfile -lopus -lm```

`-logg` should be after `-lvorbis` since vorbis needs it.

**Comment by waruqi** [comment]
> I think there's a link order problem on Linux:
> 
> ```
> > /usr/bin/g++ -o /dev/shm/.xmake1001/201212/_070815D3FE9144308463A4D2F4A6F940.b /dev/shm/.xmake1001/201212/_AF56D453AFBF4800833DFEDB0B54E920.o -m64 -L/home/runner/.xmake/packages/l/libsndfile/1.0.30/531f88cff3a64da0aa3fb05ee281ab7d/lib -L/home/runner/.xmake/packages/l/libflac/1.3.3/531f88cff3a64da0aa3fb05ee281ab7d/lib -L/home/runner/.xmake/packages/l/libogg/1.3.4/531f88cff3a64da0aa3fb05ee281ab7d/lib -L/home/runner/.xmake/packages/l/libvorbis/1.3.7/531f88cff3a64da0aa3fb05ee281ab7d/lib -L/home/runner/.xmake/packages/l/libopus/1.3.1/531f88cff3a64da0aa3fb05ee281ab7d/lib -lsndfile -lFLAC -logg -lvorbisenc -lvorbis -lvorbisfile -lopus -lm
> ```
> 
> `-logg` should be after `-lvorbis` since vorbis needs it.

You can modify deps order

```
    add_deps("cmake", "libflac", "libopus", "libvorbis", "libogg")
```

**Comment by SirLynix** [comment]
> You can modify deps order
> 
> ```
>     add_deps("cmake", "libflac", "libopus", "libvorbis", "libogg")
> ```

Yes but since libvorbis does have libogg as a dependency I assume there's an issue with xmake library ordering. I'll try this however in the meantime.

**Comment by SirLynix** [comment]
On Windows libsndfile tries to import a libFLAC function in static mode, which it shouldn't thanks to the `FLAC__NO_DLL` define.

And it's part of the command line.

```
> "C:\\Program Files (x86)\\Microsoft Visual Studio\\2019\\Enterprise\\VC\\Tools\\MSVC\\14.28.29333\\bin\\HostX64\\x64\\cl.exe" -c -nologo -DFLAC__NO_DLL -experimental:external -external:W0 -external:IC:\Users\runneradmin\AppData\Local\.xmake\packages\l\libsndfile\1.0.30\7001040a8533492597b595810d0adc4e\include -external:IC:\Users\runneradmin\AppData\Local\.xmake\packages\l\libvorbis\1.3.7\7001040a8533492597b595810d0adc4e\include -external:IC:\Users\runneradmin\AppData\Local\.xmake\packages\l\libogg\1.3.4\7001040a8533492597b595810d0adc4e\include -external:IC:\Users\runneradmin\AppData\Local\.xmake\packages\l\libopus\1.3.1\7001040a8533492597b595810d0adc4e\include -external:IC:\Users\runneradmin\AppData\Local\.xmake\packages\l\libflac\1.3.3\7001040a8533492597b595810d0adc4e\include -FoC:\Users\RUNNER~1\AppData\Local\Temp\.xmake\201213\_B5A4A450ECD94B508B56EDFC53FF49A0.o C:\Users\RUNNER~1\AppData\Local\Temp\.xmake\201213\_DFF1C46F62694B508F6CF21A725B0460.c
```

Can't figure out why it tries to import it then.

**Comment by SirLynix** [comment]
(libogg) I though maybe autoconf would be better for macos, android and such but no.

**Comment by SirLynix** [comment]
Superseded by #213

**Pull request closed without merging**